x/twap/strategy_test.go

```diff
@@ -0,0 +1,244 @@
+			quoteAsset:  denom0,
+			expTwap:     sdk.NewDec(20),
+		},
+		"denom1 start accumulator of 10 * 1s, end accumulator 10 * 1s + 20 * 2s = 20": {
+			startRecord: newOneSidedGeometricRecord(baseTime, geometricTenSecAccum),
+			endRecord:   newOneSidedGeometricRecord(baseTime.Add(time.Second*2), geometricTenSecAccum.Add(OneSec.MulInt64(2).Mul(twap.TwapLog(sdk.NewDec(20))))),
+			quoteAsset:  denom1,
+			expTwap:     sdk.OneDec().Quo(sdk.NewDec(20)),
+		},
+
+		// toggle time delta.
+		"accumulator = log(10)*OneSec, t=5s. 0 base accum": geometricTestCaseFromDeltas0(
+			s, sdk.ZeroDec(), geometricTenSecAccum, 5*time.Second, twap.TwapPow(geometricTenSecAccum.QuoInt64(5*1000))),
+		"accumulator = log(10)*OneSec, t=3s. 0 base accum": geometricTestCaseFromDeltas0(
+			s, sdk.ZeroDec(), geometricTenSecAccum, 3*time.Second, twap.TwapPow(geometricTenSecAccum.QuoInt64(3*1000))),
+		"accumulator = log(10)*OneSec, t=100s. 0 base accum": geometricTestCaseFromDeltas0(
+			s, sdk.ZeroDec(), geometricTenSecAccum, 100*time.Second, twap.TwapPow(geometricTenSecAccum.QuoInt64(100*1000))),
+
+		// test that base accum has no impact
+		"accumulator = log(10)*OneSec, t=5s. 10 base accum": geometricTestCaseFromDeltas0(
+			s, logTen, geometricTenSecAccum, 5*time.Second, twap.TwapPow(geometricTenSecAccum.QuoInt64(5*1000))),
+		"accumulator = log(10)*OneSec, t=3s. 10*second base accum": geometricTestCaseFromDeltas0(
+			s, OneSec.MulInt64(10).Mul(logTen), geometricTenSecAccum, 3*time.Second, twap.TwapPow(geometricTenSecAccum.QuoInt64(3*1000))),
+		"accumulator = 10*OneSec, t=100s. .1*second base accum": geometricTestCaseFromDeltas0(
+			s, OneSec.MulInt64(10).Mul(logOneOverTen), geometricTenSecAccum, 100*time.Second, twap.TwapPow(geometricTenSecAccum.QuoInt64(100*1000))),
+
+		// TODO: this is the highest price we currently support with the given precision bounds.
+		// Need to choose better base and potentially improve math functions to mitigate.
+		"price of 1_000_000 for an hour": {
+			startRecord: newOneSidedGeometricRecord(baseTime, sdk.ZeroDec()),
+			endRecord:   newOneSidedGeometricRecord(baseTime.Add(time.Hour), OneSec.MulInt64(60*60).Mul(twap.TwapLog(sdk.NewDec(1_000_000)))),
+			quoteAsset:  denom0,
+			expTwap:     sdk.NewDec(1_000_000),
+		},
+		// TODO: overflow tests
+		// - max spot price
+		// - large time delta
+		// - both
+
+		// TODO: hand calculated tests
+	}
+
+	for name, tc := range tests {
+		s.Run(name, func() {
+			osmoassert.ConditionalPanic(s.T(), tc.expPanic, func() {
+				geometricStrategy := &twap.GeometricTwapStrategy{TwapKeeper: *s.App.TwapKeeper}
+				actualTwap := geometricStrategy.ComputeTwap(tc.startRecord, tc.endRecord, tc.quoteAsset)
+				osmoassert.DecApproxEq(s.T(), tc.expTwap, actualTwap, osmomath.GetPowPrecision())
+			})
+		})
+	}
+}
+
+// TODO: split up this test case to cover both arithmetic and geometric twap
+func (s *TestSuite) TestComputeArithmeticTwap_ThreeAsset() {
```

Comment:
The test case is outdated. Please merge `main` or rebase.

There should be a arithmetic version: https://github.com/osmosis-labs/osmosis/blob/793bb50bdf0083f74fcb4c2c47851eb85bad91a2/x/twap/logic_test.go#L747

and a geometric version: https://github.com/osmosis-labs/osmosis/blob/793bb50bdf0083f74fcb4c2c47851eb85bad91a2/x/twap/logic_test.go#L783